contracts/WooCrossChainRouter.sol

```diff
@@ -1,372 +1,446 @@
+        bytes memory payload = _encodePayload(refId, to, dstInfos);
+
+        IStargateRouter.lzTxObj memory obj = _getLzTxObj(to, dstInfos);
+
+        return
+            stargateRouter.quoteLayerZeroFee(
+                dstInfos.chainId,
+                1, // https://stargateprotocol.gitbook.io/stargate/developers/function-types
+                obj.dstNativeAddr,
+                payload,
+                obj
+            );
+    }
+
+    function allDirectBridgeTokens() external view override returns (address[] memory) {
+        uint256 length = directBridgeTokens.length();
+        address[] memory tokens = new address[](length);
+        unchecked {
+            for (uint256 i = 0; i < length; ++i) {
+                tokens[i] = directBridgeTokens.at(i);
+            }
+        }
+        return tokens;
+    }
+
+    function allDirectBridgeTokensLength() external view override returns (uint256) {
+        return directBridgeTokens.length();
+    }
+
+    function _initSGETHs() internal {
+        // Ethereum
+        sgETHs[101] = 0x72E2F4830b9E45d52F80aC08CB2bEC0FeF72eD9c;
+        // Arbitrum
+        sgETHs[110] = 0x82CbeCF39bEe528B5476FE6d1550af59a9dB6Fc0;
+        // Optimism
+        sgETHs[111] = 0xb69c8CBCD90A39D8D3d3ccf0a3E968511C3856A0;
+    }
+
+    function _initSGPoolIds() internal {
+        // poolId > 0 means able to be bridge token
+        // Ethereum
+        sgPoolIds[101][0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48] = 1; // USDC
+        sgPoolIds[101][ETH_PLACEHOLDER_ADDR] = 13; // ETH
+        // BNB Chain
+        sgPoolIds[102][0x55d398326f99059fF775485246999027B3197955] = 2; // USDT
+        sgPoolIds[102][0xe9e7CEA3DedcA5984780Bafc599bD69ADd087D56] = 5; // BUSD
+        // Avalanche
+        sgPoolIds[106][0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E] = 1; // USDC
+        sgPoolIds[106][0x9702230A8Ea53601f5cD2dc00fDBc13d4dF4A8c7] = 2; // USDT
+        // Polygon
+        sgPoolIds[109][0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174] = 1; // USDC
+        sgPoolIds[109][0xc2132D05D31c914a87C6611C10748AEb04B58e8F] = 2; // USDT
+        // Arbitrum
+        sgPoolIds[110][0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8] = 1; // USDC
+        sgPoolIds[110][0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9] = 2; // USDT
+        sgPoolIds[110][ETH_PLACEHOLDER_ADDR] = 13; // ETH
+        // Optimism
+        sgPoolIds[111][0x7F5c764cBc14f9669B88837ca1490cCa17c31607] = 1; // USDC
+        sgPoolIds[111][ETH_PLACEHOLDER_ADDR] = 13; // ETH
+        // Fantom
+        sgPoolIds[112][0x04068DA6C83AFCFA0e13ba15A6696662335D5B75] = 1; // USDC
+    }
+
+    function _getLzTxObj(address to, DstInfos memory dstInfos) internal view returns (IStargateRouter.lzTxObj memory) {
+        uint256 dstGasForCall = (dstInfos.toToken == dstInfos.bridgeToken) ? dstGasForNoSwapCall : dstGasForSwapCall;
+
+        return IStargateRouter.lzTxObj(dstGasForCall, dstInfos.airdropNativeAmount, abi.encodePacked(to));
+    }
+
+    function _encodePayload(
```

Comment:
if this function is only one line, probably won't need to be defined as an internal function, right?

Reply:
Right, now two functions `crossSwap` & `quoteLayerZeroFee` have same encode logic, defined as internal function just for reduce the code duplicate.